benchmark/mnist_benchmark.py

```diff
@@ -0,0 +1,668 @@
+    :return: A dictionary containing the final test loss and accuracy after training.
+    """
+    model = MLP(hidden_size=64)
+
+    # Access the values from the data_bundle dictionary
+    data = data_bundle["data"]
+    targets = data_bundle["targets"]
+    test_data = data_bundle["test_data"]
+    test_targets = data_bundle["test_targets"]
+
+    result = train_and_evaluate(
+        DataSet(data, targets),
+        DataSet(test_data, test_targets),
+        model,
+        key,
+        config,
+    )
+
+    return result
+
+
+def save_results(results: dict) -> None:
+    """
+    Save benchmark results to a JSON file for algorithm performance visualisation.
+
+    :param results: A dictionary of results structured as follows:
+                    {
+                        "algorithm_name": {
+                            "coreset_size_1": {
+                                "run_1": accuracy_value,
+                                "run_2": accuracy_value,
+                                ...
+                            },
+                            "coreset_size_2": {
+                                "run_1": accuracy_value,
+                                "run_2": accuracy_value,
+                                ...
+                            },
+                            ...
+                        },
+                        "another_algorithm_name": {
+                            "coreset_size_1": {
+                                "run_1": accuracy_value,
+                                "run_2": accuracy_value,
+                                ...
+                            },
+                            ...
+                        },
+                        ...
+                    }
+                    Each algorithm contains coreset sizes as keys, with values being
+                    dictionaries of accuracy results from different runs.
+    """
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+    with open(
+        os.path.join(base_dir, "mnist_benchmark_results.json"), "w", encoding="utf-8"
+    ) as f:
+        json.dump(results, f, indent=4)
+
+    print("Data has been saved to 'benchmark_results.json'")
+
+
+def main() -> None:
+    """
+    Perform the benchmark for multiple solvers, coreset sizes, and random seeds.
+
+    The function follows these steps:
+    1. Prepare and load the MNIST datasets (training and test).
+    2. Perform dimensionality reduction on the training data using PCA.
+    3. Initialise solvers for data reduction.
+    4. For each solver and coreset size, reduce the dataset and train the model
+       on the reduced set.
+    5. Train the model and evaluate its performance on the test set.
+    6. Save the results, which include test accuracy for each solver and coreset size.
+    """
+    (train_data_jax, train_targets_jax, test_data_jax, test_targets_jax) = (
+        prepare_datasets()
+    )
+
+    train_data_pca = pca(train_data_jax)
```

Comment:
Consider converting to `Data` object here, rather than in the nested for-loop, so that we're not doing this over and over below?

`train_data_pca = Data(pca(train_data_jax))`